File: go/internal/core/ndjson.go
package core

import (
	"encoding/json"
)

// Command represents a matching engine command.
type Command struct {
	PlaceSell    *PlaceSellCmd    `json:"PlaceSell,omitempty"`
	CancelSell   *CancelSellCmd   `json:"CancelSell,omitempty"`
	BuyByQty     *BuyByQtyCmd     `json:"BuyByQty,omitempty"`
	BuyByBudget  *BuyByBudgetCmd  `json:"BuyByBudget,omitempty"`
}

type PlaceSellCmd struct {
	ID        ID        `json:"id"`
	Price     Price     `json:"price"`
	Qty       AssetQty  `json:"qty"`
	Timestamp Timestamp `json:"timestamp"`
}

type CancelSellCmd struct {
	ID ID `json:"id"`
}

type BuyByQtyCmd struct {
	ID        ID        `json:"id"`
	Qty       AssetQty  `json:"qty"`
	Timestamp Timestamp `json:"timestamp"`
}

type BuyByBudgetCmd struct {
	ID        ID        `json:"id"`
	Budget    Money     `json:"budget"`
	Timestamp Timestamp `json:"timestamp"`
}

// Event represents a matching engine event.
type Event struct {
	Accepted       *struct{}       `json:"Accepted,omitempty"`
	Rejected       *RejectedEvent  `json:"Rejected,omitempty"`
	Trade          *TradeEvent     `json:"Trade,omitempty"`
	SellUpdated    *struct{}       `json:"SellUpdated,omitempty"`
	SellClosed     *struct{}       `json:"SellClosed,omitempty"`
	BuyResultQty   *BuyResultQty   `json:"BuyResultQty,omitempty"`
	BuyResultBudget *BuyResultBudget `json:"BuyResultBudget,omitempty"`
}

type RejectedEvent struct {
	Reason string `json:"reason"`
}

type TradeEvent struct {
	BuyerID  ID        `json:"buyer_id"`
	SellerID ID        `json:"seller_id"`
	Qty      AssetQty  `json:"qty"`
	Price    Price     `json:"price"`
}

type BuyResultQty struct {
	Filled AssetQty `json:"filled"`
}

type BuyResultBudget struct {
	Spent  Money    `json:"spent"`
	Filled AssetQty `json:"filled"`
}

// ParseCommand parses a JSON line into a Command.
func ParseCommand(s string) (*Command, error) {
	var cmd Command
	err := json.Unmarshal([]byte(s), &cmd)
	if err != nil {
		return nil, err
	}
	return &cmd, nil
}

// SerializeEvent serializes an Event to JSON.
func SerializeEvent(e *Event) (string, error) {
	b, err := json.Marshal(e)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

File: rust/crates/engine_core/src/ndjson.rs
//! Command and Event types for NDJSON I/O.

use serde::{Deserialize, Serialize};

use crate::types::{AssetQty, ID, Money, Price, Timestamp};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Command {
    PlaceSell {
        id: ID,
        price: Price,
        qty: AssetQty,
        timestamp: Timestamp,
    },
    CancelSell { id: ID },
    BuyByQty {
        id: ID,
        qty: AssetQty,
        timestamp: Timestamp,
    },
    BuyByBudget {
        id: ID,
        budget: Money,
        timestamp: Timestamp,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Event {
    Accepted,
    Rejected { reason: String },
    Trade {
        buyer_id: ID,
        seller_id: ID,
        qty: AssetQty,
        price: Price,
    },
    SellUpdated,
    SellClosed,
    BuyResultQty { filled: AssetQty },
    BuyResultBudget { spent: Money, filled: AssetQty },
}

/// Parse a JSON line into a Command.
pub fn parse_command(s: &str) -> Result<Command, serde_json::Error> {
    serde_json::from_str(s)
}

/// Serialize an Event to JSON.
pub fn serialize_event(event: &Event) -> String {
    serde_json::to_string(event).unwrap()
}
